README.md
# wepdl (Waveshare Electronic Paper Display Library)
(repository transfered from old account - tmseth)

wepdl currently support only UART display 4.3 inch (800x600 px). This library requires no other dependencies, has been simplified at the same time it has more options (like pin addressing in the project code instead of the library).

All informations on Wiki page : [https://github.com/8tm/wepdl/wiki]


<img src="wepdl_screen.jpg" />

---

## Running the tests

Select menu File -> Examples -> wepdl -> (some example)

1) Connect your device (Arduino, Nodemcu, etc) to computer.

2) Choice USB port.

3) In Arduino-IDE use combination CTRL + U to compile and upload sketch.

4) Wait a moment :)

---

## Authors

* **Tadeusz Miszczyk** - *This EPD library*
* **Waveshare** - *Initial work with oryginal EPD library*

---

## Waveshare

Website : [https://www.waveshare.com](https://www.waveshare.com)

E-mail : [service@waveshare.com](mailto:service@waveshare.com)

---

Waveshare demo video on YouTube : [https://www.youtube.com/watch?v=06rDuQzYovk](https://www.youtube.com/watch?v=06rDuQzYovk)

Waveshare 4.3" e-paper display UART module : [https://www.waveshare.com/4.3inch-e-paper.htm](https://www.waveshare.com/4.3inch-e-paper.htm) (800x600 px, 4 color)

Wiki :  [https://www.waveshare.com/wiki/4.3inch_e-Paper_UART_Module](https://www.waveshare.com/wiki/4.3inch_e-Paper_UART_Module)

Datasheets of parts : [https://www.waveshare.com/wiki/4.3inch_e-Paper_Datasheets](https://www.waveshare.com/wiki/4.3inch_e-Paper_Datasheets)

Oryginal library : [https://www.waveshare.com/w/upload/archive/e/eb/20171228082252%214.3inch-e-Paper-Code.7z](https://www.waveshare.com/w/upload/archive/e/eb/20171228082252%214.3inch-e-Paper-Code.7z)

User manual : [https://www.waveshare.com/wiki/File:4.3inch-e-Paper-UserManual.pdf](https://www.waveshare.com/wiki/File:4.3inch-e-Paper-UserManual.pdf)

---

## License

This project is licensed under the GNU General Public License v3.0 - see the [LICENSE](LICENSE) file for details

---

## Acknowledgments

* Inspiration : Life

---


changelog.md
## 0.0.3 - 2019/11/06
- Changed folder structure (matching the folder structure to the one generally adopted in Arduino libraries)
- Changed semantic versioning convention from 2 segments to 3 segments (This change may mislead, so it is better
  to acknowledge that previous versions did not exist and the project started from the current version 0.0.3)

# ======================================== OLD SEMANTIC VERSIONING =============================================
## 0.2 - 2018/04/18
- Repository transfered to new account 8tm (from tmseth)

## 0.2 - 2018/04/18
- Changed directory structure
- Added new full examples
- Moved (some deleted) informations from README.md to Wiki pages

## 0.2 - 2018/04/17
- Added informations from README.md to Wiki pages

## 0.2 - 2018/01/06
- Changed names of variables, methods etc
- Added to method drawLine option "thickness" - default is set to 1px
- Changed wepdl screen
- Updated changelog
- Updated REAMDE.md
- Added datasheets of parts and firmware for 4.3" display


## 0.1 - 2018/01/06
- First "dirty" release
- Created from oryginal Waveshare EPD library
- One class "wepdl" (for now)
- While creating object you have to set pins for RX, TX, WAKEP, RESET (You don't have to do it in library from now)
- Added draw method for transparent Rectangle
- Added draw method to create "ring" (not transparent)
- Tested (and works) with Waveshare 4.3" e-ink display (UART) on Lolin NodeMcu v3, Adruino Uno and Arduino Nano
- Created .gitignore file

## 0.0 - 2018/01/05
- Initial commit
- Created README.md, LICENSE file


wepdl.h
// =====================================================================================================================
//
//    File             :  wepdl.h
//    Tested on        :  Arduino-IDE (1.8.2 & 1.9.0 Beta31), Arduino Uno/Nano; Lolin Nodemcu v3, Wemos d1 mini
//    Version          :  0.0.3
//    Created by       :  Tadeusz Miszczyk (tadeusz.miszczyk[at]gmail.com)
//    Library          :  https://github.com/8tm/wepdl
//    Wiki             :  https://github.com/8tm/wepdl/wiki
//    Last changes     :  2019-11-06
//
// =====================================================================================================================
#ifndef    WEPDL_H
#define    WEPDL_H

#include <Arduino.h>
#include <SoftwareSerial.h>

// Memory
#define    FLASH          0
#define    MICROSD        1

// Screen rotation
#define    HORIZONTAL     0
#define    VERTICAL       1

// Languages
#define    ENGLISH        0x1E
#define    CHINESE        0x1F

// Font sizes
#define    SIZE32         0x01
#define    SIZE48         0x02
#define    SIZE64         0x03

// Colors
#define    BLACK          0x00
#define    DARK           0x01
#define    LIGHT          0x02
#define    WHITE          0x03

//----------------------------------------------------------------------------------------------------------------------
class wepdl
{
    private :
            uint8_t             _wake_up,
                                _reset;

            unsigned char       _global_color,
                                _global_background_color,
                                _cmd_buff[512];

            const unsigned char _cmd_handshake[8] = { 0xA5, 0x00, 0x09, 0x00, 0xCC, 0x33, 0xC3, 0x3C },
                                _cmd_read_baud[8] = { 0xA5, 0x00, 0x09, 0x02, 0xCC, 0x33, 0xC3, 0x3C },
                                _cmd_stopmode[8]  = { 0xA5, 0x00, 0x09, 0x08, 0xCC, 0x33, 0xC3, 0x3C },
                                _cmd_update[8]    = { 0xA5, 0x00, 0x09, 0x0A, 0xCC, 0x33, 0xC3, 0x3C },
                                _cmd_load_font[8] = { 0xA5, 0x00, 0x09, 0x0E, 0xCC, 0x33, 0xC3, 0x3C },
                                _cmd_load_pic[8]  = { 0xA5, 0x00, 0x09, 0x0F, 0xCC, 0x33, 0xC3, 0x3C };

            void _putchars       ( const unsigned char * ptr, int n);
            unsigned char _verify( const void * ptr, int n);


    public :
            wepdl                ( uint8_t rx,
                                   uint8_t tx,
                                   uint8_t wake_up,
                                   uint8_t reset );

            SoftwareSerial        *SoftSerial;

            // ------------- D E V I C E -------------------------------------------------------------------------------------------
            void initialize      ( void );
            void goSleep         ( void );
            void handShake       ( void );
            void reset           ( void );
            void wakeUp          ( void );
            void readBaud        ( void );

            // ------------- S C R E E N -------------------------------------------------------------------------------------------
            void clearScreen     ( void );
            void rotateScreen    ( unsigned char mode );
            void updateScreen    ( void );

            // ------------- S E T -------------------------------------------------------------------------------------------------
            void setBaud         ( long baud );
            void setColor        ( unsigned char color,
                                   unsigned char background_color,
                                   bool global_color = true );

            void setFont         ( unsigned char font,
                                   unsigned char language = ENGLISH );

            void setMemory       ( unsigned char mode );

            // ------------- D I S P L A Y -----------------------------------------------------------------------------------------
            void displayBitmap   ( int x,
                                   int y,
                                   const void * p );

            void displayCharacter( int x,
                                   int y,
                                   unsigned char ch );

            void displayText     ( int x,
                                   int y,
                                   const void * p );

            // ------------- I M A G E S -------------------------------------------------------------------------------------------

            void Picture         ( void );

            // ------------- ? ? ? ? -----------------------------------------------------------------------------------------------
            void FontLoad        ( void );

            // ------------- D R A W -----------------------------------------------------------------------------------------------
            void drawPixel       ( int x,
                                   int y );

            void drawLine        ( int A_x,
                                   int A_y,
                                   int B_x,
                                   int B_y,
                                   int thickness = 1 );

            void drawCircle      ( int x,
                                   int y,
                                   int radius,
                                   bool fill = false );

            void drawRing        ( int x,
                                   int y,
                                   int radius,
                                   int border,
                                   unsigned char color,
                                   unsigned char background_color );

            void drawTriangle    ( int A_x,
                                   int A_y,
                                   int B_x,
                                   int B_y,
                                   int C_x,
                                   int C_y,
                                   bool fill = false );

            void drawRectangle   ( int A_x,
                                   int A_y,
                                   int C_x,
                                   int C_y,
                                   bool fill = false );

};
//----------------------------------------------------------------------------------------------------------------------
#endif



wepdl.cpp
// =====================================================================================================================
//
//    File             :  wepdl.cpp
//    Tested on        :  Arduino-IDE (1.8.2 & 1.9.0 Beta31), Arduino Uno/Nano; Lolin Nodemcu v3, Wemos d1 mini
//    Version          :  0.0.3
//    Created by       :  Tadeusz Miszczyk (tadeusz.miszczyk[at]gmail.com)
//    Library          :  https://github.com/8tm/wepdl
//    Wiki             :  https://github.com/8tm/wepdl/wiki
//    Last changes     :  2019-11-06
//
// =====================================================================================================================
#include "wepdl.h"
//----------------------------------------------------------------------------------------------------------------------
extern "C" {
#include "user_interface.h"
}
//----------------------------------------------------------------------------------------------------------------------
wepdl :: wepdl ( uint8_t rx,
                 uint8_t tx,
                 uint8_t wake_up,
                 uint8_t reset )
{
    _wake_up   = wake_up;
    _reset     = reset;
    SoftSerial = new SoftwareSerial( rx, tx );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::_putchars( const unsigned char * ptr,
                       int n )
{
    int i,
        x;

    for( i = 0; i < n; i++ )
    {
        x = ptr[i];
        SoftSerial->write( x );
    }
}
//----------------------------------------------------------------------------------------------------------------------
unsigned char wepdl::_verify( const void * ptr,
                              int n )
{
    int i;
    unsigned char * p = ( unsigned char * ) ptr;
    unsigned char result;
    for( i = 0, result = 0; i < n; i++ )
    {
        result ^= p[i];
    }
    return result;
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::initialize ( void )
{
    SoftSerial->begin( 115200 );

    pinMode( _wake_up, HIGH );

    pinMode( _reset, HIGH );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::goSleep       ( void )
{
    memcpy( _cmd_buff, _cmd_stopmode, 8 );

    _cmd_buff[8] = _verify( _cmd_buff, 8 );

    _putchars( _cmd_buff, 9 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::handShake  ( void )
{
    memcpy( _cmd_buff, _cmd_handshake, 8 );

    _cmd_buff[8] = _verify( _cmd_buff, 8 );

    _putchars( _cmd_buff, 9 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::reset( void )
{
    digitalWrite( _reset, LOW );
    delayMicroseconds( 10 );

    digitalWrite( _reset, HIGH );
    delayMicroseconds( 500 );

    digitalWrite( _reset, LOW );
    delay( 3000 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::wakeUp     ( void )
{
    digitalWrite( _wake_up, LOW );
    delayMicroseconds( 10 );

    digitalWrite( _wake_up, HIGH );
    delayMicroseconds( 500 );

    digitalWrite( _wake_up, LOW );
    delay( 10 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::readBaud   ( void )
{
    memcpy( _cmd_buff, _cmd_read_baud, 8 );

    _cmd_buff[8] = _verify( _cmd_buff, 8 );

    _putchars( _cmd_buff, 9 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::clearScreen    ( void )
{
    _cmd_buff[0] = 0xA5;
    _cmd_buff[1] = 0x00;
    _cmd_buff[2] = 0x09;
    _cmd_buff[3] = 0x2E;
    _cmd_buff[4] = 0xCC;
    _cmd_buff[5] = 0x33;
    _cmd_buff[6] = 0xC3;
    _cmd_buff[7] = 0x3C;
    _cmd_buff[8] = _verify( _cmd_buff, 8 );

    _putchars( _cmd_buff, 9 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::rotateScreen ( unsigned char mode )
{
    _cmd_buff[0] = 0xA5;
    _cmd_buff[1] = 0x00;
    _cmd_buff[2] = 0x0A;
    _cmd_buff[3] = 0x0D;
    _cmd_buff[4] = mode;
    _cmd_buff[5] = 0xCC;
    _cmd_buff[6] = 0x33;
    _cmd_buff[7] = 0xC3;
    _cmd_buff[8] = 0x3C;
    _cmd_buff[9] = _verify( _cmd_buff, 9 );

    _putchars( _cmd_buff, 10 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::updateScreen   ( void )
{
    memcpy( _cmd_buff, _cmd_update, 8 );

    _cmd_buff[8] = _verify( _cmd_buff, 8 );

    _putchars( _cmd_buff, 9 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::setBaud    ( long baud )
{
    _cmd_buff[0]  = 0xA5;
    _cmd_buff[1]  = 0x00;
    _cmd_buff[2]  = 0x0D;
    _cmd_buff[3]  = 0x01;

    _cmd_buff[4]  = ( baud >> 24 ) & 0xFF;
    _cmd_buff[5]  = ( baud >> 16 ) & 0xFF;
    _cmd_buff[6]  = ( baud >> 8  ) & 0xFF;
    _cmd_buff[7]  =   baud & 0xFF;

    _cmd_buff[8]  = 0xCC;
    _cmd_buff[9]  = 0x33;
    _cmd_buff[10] = 0xC3;
    _cmd_buff[11] = 0x3C;
    _cmd_buff[12] = _verify( _cmd_buff, 12 );

    _putchars( _cmd_buff, 13 );
    delay( 10 );
    SoftSerial->begin( baud );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::setColor ( unsigned char color,
                       unsigned char background_color,
                       bool global_color )
{
    _cmd_buff[0]  = 0xA5;
    _cmd_buff[1]  = 0x00;
    _cmd_buff[2]  = 0x0B;
    _cmd_buff[3]  = 0x10;
    _cmd_buff[4]  = color;
    _cmd_buff[5]  = background_color;
    _cmd_buff[6]  = 0xCC;
    _cmd_buff[7]  = 0x33;
    _cmd_buff[8]  = 0xC3;
    _cmd_buff[9]  = 0x3C;
    _cmd_buff[10] = _verify( _cmd_buff, 10 );

    _putchars( _cmd_buff, 11 );

    if ( global_color )
    {
        _global_color = color;
        _global_background_color = background_color;
    }
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::setFont  ( unsigned char font,
                       unsigned char language )
{
    _cmd_buff[0] = 0xA5;
    _cmd_buff[1] = 0x00;
    _cmd_buff[2] = 0x0A;
    _cmd_buff[3] = language;
    _cmd_buff[4] = font;
    _cmd_buff[5] = 0xCC;
    _cmd_buff[6] = 0x33;
    _cmd_buff[7] = 0xC3;
    _cmd_buff[8] = 0x3C;
    _cmd_buff[9] = _verify( _cmd_buff, 9 );

    _putchars( _cmd_buff, 10 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::setMemory     ( unsigned char mode )
{
    _cmd_buff[0] = 0xA5;
    _cmd_buff[1] = 0x00;
    _cmd_buff[2] = 0x0A;
    _cmd_buff[3] = 0x07;
    _cmd_buff[4] = mode;
    _cmd_buff[5] = 0xCC;
    _cmd_buff[6] = 0x33;
    _cmd_buff[7] = 0xC3;
    _cmd_buff[8] = 0x3C;
    _cmd_buff[9] = _verify( _cmd_buff, 9 );

    _putchars( _cmd_buff, 10 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::displayBitmap  ( int x,
                             int y,
                             const void * p )
{
    int string_size;
    unsigned char * ptr = ( unsigned char * ) p;

    string_size  = strlen( ( const char * ) ptr );
    string_size += 14;

    _cmd_buff[0] = 0xA5;
    _cmd_buff[1] = ( string_size >> 8 ) & 0xFF;
    _cmd_buff[2] =   string_size & 0xFF;
    _cmd_buff[3] = 0x70;
    _cmd_buff[4] = ( x >> 8 ) & 0xFF;
    _cmd_buff[5] =   x & 0xFF;
    _cmd_buff[6] = ( y >> 8 ) & 0xFF;
    _cmd_buff[7] =   y & 0xFF;

    strcpy( ( char * )( &_cmd_buff[8] ), ( const char * ) ptr );

    string_size -= 5;

    _cmd_buff[string_size]     = 0xCC;
    _cmd_buff[string_size + 1] = 0x33;
    _cmd_buff[string_size + 2] = 0xC3;
    _cmd_buff[string_size + 3] = 0x3C;
    _cmd_buff[string_size + 4] = _verify( _cmd_buff, string_size + 4 );

    _putchars( _cmd_buff, string_size + 5 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::displayCharacter ( int x,
                               int y,
                               unsigned char ch )
{
    unsigned char buff[2];
    buff[0] = ch;
    buff[1] = 0;
    displayText( x, y, buff );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::displayText ( int x,
                          int y,
                          const void * p )
{
    int string_size;
    unsigned char * ptr = ( unsigned char * ) p;

    string_size = strlen( ( const char * ) ptr );
    string_size += 14;

    _cmd_buff[0] = 0xA5;

    _cmd_buff[1] = ( string_size >> 8 ) & 0xFF;
    _cmd_buff[2] =   string_size & 0xFF;

    _cmd_buff[3] = 0x30;

    _cmd_buff[4] = ( x >> 8 ) & 0xFF;
    _cmd_buff[5] =   x & 0xFF;
    _cmd_buff[6] = ( y >> 8 ) & 0xFF;
    _cmd_buff[7] =   y & 0xFF;

    strcpy( ( char * )( &_cmd_buff[8] ), ( const char * ) ptr );

    string_size -= 5;

    _cmd_buff[string_size]     = 0xCC;
    _cmd_buff[string_size + 1] = 0x33;
    _cmd_buff[string_size + 2] = 0xC3;
    _cmd_buff[string_size + 3] = 0x3C;
    _cmd_buff[string_size + 4] = _verify( _cmd_buff, string_size + 4 );
    _putchars( _cmd_buff, string_size + 5 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::Picture ( void )
{
    memcpy( _cmd_buff, _cmd_load_pic, 8 );

    _cmd_buff[8] = _verify( _cmd_buff, 8 );

    _putchars( _cmd_buff, 9 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::FontLoad    ( void )
{
    memcpy( _cmd_buff, _cmd_load_font, 8 );

    _cmd_buff[8] = _verify( _cmd_buff, 8 );

    _putchars( _cmd_buff, 9 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::drawPixel     ( int x,
                            int y )
{
    _cmd_buff[0]  = 0xA5;
    _cmd_buff[1]  = 0x00;
    _cmd_buff[2]  = 0x0D;
    _cmd_buff[3]  = 0x20;
    _cmd_buff[4]  = ( x >> 8 ) & 0xFF;
    _cmd_buff[5]  =   x & 0xFF;
    _cmd_buff[6]  = ( y >> 8 ) & 0xFF;
    _cmd_buff[7]  =   y & 0xFF;
    _cmd_buff[8]  = 0xCC;
    _cmd_buff[9]  = 0x33;
    _cmd_buff[10] = 0xC3;
    _cmd_buff[11] = 0x3C;
    _cmd_buff[12] = _verify( _cmd_buff, 12 );

    _putchars( _cmd_buff, 13 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::drawLine      ( int A_x,
                            int A_y,
                            int B_x,
                            int B_y,
                            int thickness )
{
    int counter = 1,
        Ax      = A_x,
        Ay      = A_y,
        Bx      = B_x,
        By      = B_y;

    for ( counter = 1; counter <= thickness; counter++ )
    {
        _cmd_buff[0]  = 0xA5;
        _cmd_buff[1]  = 0x00;
        _cmd_buff[2]  = 0x11;
        _cmd_buff[3]  = 0x22;
        _cmd_buff[4]  = ( Ax >> 8 ) & 0xFF;
        _cmd_buff[5]  =   Ax & 0xFF;
        _cmd_buff[6]  = ( Ay >> 8 ) & 0xFF;
        _cmd_buff[7]  =   Ay & 0xFF;
        _cmd_buff[8]  = ( Bx >> 8 ) & 0xFF;
        _cmd_buff[9]  =   Bx & 0xFF;
        _cmd_buff[10] = ( By >> 8 ) & 0xFF;
        _cmd_buff[11] =   By & 0xFF;
        _cmd_buff[12] = 0xCC;
        _cmd_buff[13] = 0x33;
        _cmd_buff[14] = 0xC3;
        _cmd_buff[15] = 0x3C;
        _cmd_buff[16] = _verify( _cmd_buff, 16 );

        _putchars( _cmd_buff, 17 );

        if        ( A_y == B_y ) { Ay++; By++; }
        else if   ( A_x == B_x ) { Ax++; Bx++; }
        else if ( ( A_x != B_x ) && ( A_y != B_y ) )
             {
                 if      ( A_y < B_y ) { Ax++; Bx++; Ay--; By--; }
                 else if ( A_y > B_y ) { Ax++; Bx++; Ay++; By++; };
             };
    }
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::drawCircle    ( int x,
                            int y,
                            int radius,
                            bool fill )
{
    _cmd_buff[0]  = 0xA5;
    _cmd_buff[1]  = 0x00;
    _cmd_buff[2]  = 0x0F;

    if (fill) { _cmd_buff[3] = 0x27; }
    else      { _cmd_buff[3] = 0x26; };

    _cmd_buff[4]  = ( x >> 8 ) & 0xFF;
    _cmd_buff[5]  =   x & 0xFF;
    _cmd_buff[6]  = ( y >> 8 ) & 0xFF;
    _cmd_buff[7]  =   y & 0xFF;
    _cmd_buff[8]  = ( radius >> 8 ) & 0xFF;
    _cmd_buff[9]  =   radius & 0xFF;
    _cmd_buff[10] = 0xCC;
    _cmd_buff[11] = 0x33;
    _cmd_buff[12] = 0xC3;
    _cmd_buff[13] = 0x3C;
    _cmd_buff[14] = _verify( _cmd_buff, 14 );

    _putchars( _cmd_buff, 15 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::drawRing    ( int x,
                          int y,
                          int radius,
                          int border,
                          unsigned char color,
                          unsigned char background_color )
{
    setColor   ( color, background_color, false );
    drawCircle ( x, y, radius, true );

    setColor   ( background_color, color, false );
    drawCircle ( x, y, ( ( border - radius ) * -1 ), true );

    setColor   ( _global_color, _global_background_color );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::drawTriangle  ( int A_x,
                            int A_y,
                            int B_x,
                            int B_y,
                            int C_x,
                            int C_y,
                            bool fill )
{
    _cmd_buff[0]  = 0xA5;
    _cmd_buff[1]  = 0x00;
    _cmd_buff[2]  = 0x15;

      if (fill)  { _cmd_buff[3] = 0x29; }
      else       { _cmd_buff[3] = 0x28; };

    _cmd_buff[4]  = ( A_x >> 8 ) & 0xFF;
    _cmd_buff[5]  =   A_x & 0xFF;
    _cmd_buff[6]  = ( A_y >> 8 ) & 0xFF;
    _cmd_buff[7]  =   A_y & 0xFF;
    _cmd_buff[8]  = ( B_x >> 8 ) & 0xFF;
    _cmd_buff[9]  =   B_x & 0xFF;
    _cmd_buff[10] = ( B_y >> 8 ) & 0xFF;
    _cmd_buff[11] =   B_y & 0xFF;
    _cmd_buff[12] = ( C_x >> 8 ) & 0xFF;
    _cmd_buff[13] =   C_x & 0xFF;
    _cmd_buff[14] = ( C_y >> 8 ) & 0xFF;
    _cmd_buff[15] =   C_y & 0xFF;
    _cmd_buff[16] = 0xCC;
    _cmd_buff[17] = 0x33;
    _cmd_buff[18] = 0xC3;
    _cmd_buff[19] = 0x3C;
    _cmd_buff[20] = _verify( _cmd_buff, 20 );

    _putchars( _cmd_buff, 21 );
}
//----------------------------------------------------------------------------------------------------------------------
void wepdl::drawRectangle ( int A_x,
                            int A_y,
                            int C_x,
                            int C_y,
                            bool fill )
{
    _cmd_buff[0]  = 0xA5;
    _cmd_buff[1]  = 0x00;
    _cmd_buff[2]  = 0x11;

    if (fill)  { _cmd_buff[3] = 0x24; }
    else       { _cmd_buff[3] = 0x25; };

    _cmd_buff[4]  = ( A_x >> 8 ) & 0xFF;
    _cmd_buff[5]  =   A_x & 0xFF;
    _cmd_buff[6]  = ( A_y >> 8 ) & 0xFF;
    _cmd_buff[7]  =   A_y & 0xFF;
    _cmd_buff[8]  = ( C_x >> 8 ) & 0xFF;
    _cmd_buff[9]  =   C_x & 0xFF;
    _cmd_buff[10] = ( C_y >> 8 ) & 0xFF;
    _cmd_buff[11] =   C_y & 0xFF;
    _cmd_buff[12] = 0xCC;
    _cmd_buff[13] = 0x33;
    _cmd_buff[14] = 0xC3;
    _cmd_buff[15] = 0x3C;
    _cmd_buff[16] = _verify( _cmd_buff, 16 );

    _putchars( _cmd_buff, 17 );
}
//----------------------------------------------------------------------------------------------------------------------




